refactor!: MontyExperiment context manager part 2

jeremyshoemaker opened the pull request:
This is the second phase of [the work I proposed on Discourse][forum] to make MontyExperiment into a context manager.

This moves the config setup into a constructor for the class, leaving the part that initializes the model and datasets in the `setup_experiment` method. Then we call the simplified setup method in the context manager `__enter__` method.

The rest of this is updating the call sites to use the new context manager style more completely.

[forum]: https://thousandbrains.discourse.group/t/investigation-into-linux-test-failures-and-possible-improvements/468

tristanls commented on src/tbp/monty/frameworks/experiments/monty_experiment.py:
```diff
@@ -51,19 +51,25 @@ class MontyExperiment:
-    def setup_experiment(self, config):
-        """Set up the basic elements of a Monty experiment and initialize counters.
+    def __init__(self, config):
+        """Initialize the experiment based on the provide configuration.
```
**nitpick (typo)**: "provided"

tristanls replied:
**nitpick (typo)**: "provided" (I wouldn't hold up a merge because of this, but since I left other comments, just highlighting this typo again).

jeremyshoemaker replied:
That's weird, I thought I changed that.

tristanls commented on tests/unit/base_config_test.py:
```diff
@@ -79,17 +79,14 @@ def test_can_set_up(self):
-        self.exp = MontyExperiment()
-        with self.exp:
-            self.exp.setup_experiment(base_config)
+        with MontyExperiment(base_config) as self.exp:
```
**suggestion (non-blocking)**: I don't see a need to attach `exp` to `self` in these tests. `setUp` and `tearDown` do not interact with `self.exp`. I think here and in all other tests (in this file and all the others) this can be:
```python
with MontyExperiment(base_config) as exp:
```

jeremyshoemaker replied:
I didn't think so either, but I was trying to make the minimal changes necessary. I'll go ahead and change all of them to not stash it on the test instance.

tristanls commented on tests/unit/evidence_lm_test.py:
```diff
@@ -1702,9 +1683,7 @@ def test_5lm_evidence_experiment(self):
-        self.exp = MontyObjectRecognitionExperiment()
-        with self.exp:
-            self.exp.setup_experiment(config)
+        with MontyObjectRecognitionExperiment(config) as self.exp:
```
**issue**: Comment on line 1718/1697 below is outdated.
```
 # not running self.exp.evaluate here to avoid exp.close being called.
```

tristanls commented on tests/unit/graph_learning_test.py:
```diff
@@ -547,16 +547,13 @@ def test_can_set_up(self):
-        with self.exp:
-            self.exp.setup_experiment(base_config)
+        with MontyObjectRecognitionExperiment(base_config) as self.exp:
+            pass
```
**issue**: I think the test should be renamed to something like `test_can_initialize` and the docstring updated accordingly.

tristanls commented on tests/unit/graph_learning_test.py:
```diff
@@ -689,9 +674,7 @@ def test_fixed_actions_disp(self):
-        with self.exp:
-            self.exp.setup_experiment(config)
+        with MontyObjectRecognitionExperiment(config) as self.exp:
             # self.exp.model.set_experiment_mode("eval")
```
**issue (non-blocking)**: Remove commented out code.

tristanls commented on tests/unit/graph_learning_test.py:
```diff
@@ -713,9 +696,7 @@ def test_fixed_actions_ppf(self):
-        with self.exp:
-            self.exp.setup_experiment(config)
+        with MontyObjectRecognitionExperiment(config) as self.exp:
             # self.exp.model.set_experiment_mode("eval")
```
**issue (non-blocking)**: Remove commented out code.

tristanls reviewed (commented):
Thank you for the follow-up. This is a breaking change since how someone would invoke an experiment has changed. It is perfectly fine, but let's be explicit about it.

**issue**: This is a breaking change in that an experiment now must be initialized with a `config` as an initialization parameter. Please update the commit and PR prefix to `refactor!:` to indicate the breaking change.

**issue**: `ProfileExperimentMixin` calls `super().setup_experiment(config)` inside its `setup_experiment(self, config)` method and invokes `super().close()` directly. Does that code still work?

jeremyshoemaker commented:
> Thank you for the follow-up. This is a breaking change since how someone would invoke an experiment has changed. It is perfectly fine, but let's be explicit about it.
>
> **issue**: This is a breaking change in that an experiment now must be initialized with a `config` as an initialization parameter. Please update the commit and PR prefix to `refactor!:` to indicate the breaking change.

Will do. 

> **issue**: `ProfileExperimentMixin` calls `super().setup_experiment(config)` inside its `setup_experiment(self, config)` method and invokes `super().close()` directly. Does that code still work?

Good catch, I'm not sure how I missed this call site. Looking at what the code does, I'm pretty sure it will still work, except that we won't be profiling all the config setting work that I moved to the MontyExperiment constructor. It shouldn't be much from a performance perspective, and I think the profiling is really trying to capture how long it takes to setup the model and load the datasets.

All that said, we don't appear to have any test coverage for this mixin, so it might make sense for me to add some to test that this change doesn't break anything and to make sure it's covered in the future.

vkakerbeck commented on src/tbp/monty/frameworks/experiments/monty_experiment.py:
```diff
@@ -51,19 +51,25 @@ class MontyExperiment:
-        # Save a copy of the config used to specify the experiment before modifying
+        # Copy the config and store it so we can modify it freely
         config = copy.deepcopy(config)
-        # Convert any dataclass back to dict for backward compatibility
```
Just curious why you removed this comment. Is it not accurate?

jeremyshoemaker replied:
I thought it was redundant because it's basically summarizing the docstring of `config_to_dict`. 

The other comment above got rewritten because I originally copied the code up from `setup_experiment` without the comment and wrote a new one. I didn't realize I ended up just rewriting the original comment until I saw the diff. I think my version _might_ be explaining the "why" of the copying better than the original, but I could just as easily keep the original comment. 🤷

vkakerbeck replied:
okay yeah makes sense. I'm happy with how it is. I was just curious if something changed in the functionality that I missed.

vkakerbeck reviewed (approved):
Looks good, thanks for the update!

nielsleadholm reviewed (approved):
Just had a quick look through but overall looks good from my end, thanks for making this change.

jeremyshoemaker commented:
For posterity, I added unit tests for the **ProfileExperimentMixin** in #222. I'll pull those changes in here and make sure everything still works as expected.

jeremyshoemaker commented:
Rebased onto `main` to pull in the work I did in #222, and addressed PR feedback. Most of it is just changing `self.exp` to `exp` in the tests since we don't need to be saving it on the test case objects.

tristanls commented on tests/unit/graph_building_test.py:
```diff
@@ -224,23 +224,22 @@ def check_graph_formatting(self, graph, features_to_check):
             pprint("...training...")
-            self.exp.train()
+            exp.train()
+        return exp
```
**note**: TIL that `with` does not create a scope, huh.

tristanls commented on tests/unit/graph_learning_test.py:
```diff
@@ -689,98 +674,86 @@ def test_fixed_actions_disp(self):
-            self.exp.setup_experiment(config)
-            # self.exp.model.set_experiment_mode("eval")
+        with MontyObjectRecognitionExperiment(config) as exp:
+            # exp.model.set_experiment_mode("eval")
```
**issue (non-blocking)**: Remove commented out code.

tristanls commented on tests/unit/graph_learning_test.py:
```diff
@@ -689,98 +674,86 @@ def test_fixed_actions_disp(self):
-            self.exp.setup_experiment(config)
-            # self.exp.model.set_experiment_mode("eval")
+        with MontyObjectRecognitionExperiment(config) as exp:
+            # exp.model.set_experiment_mode("eval")
```
**issue (non-blocking)**: Remove commented out code.

tristanls commented on tests/unit/graph_learning_test.py:
```diff
@@ -689,98 +674,86 @@ def test_fixed_actions_disp(self):
-        self.eval_exp_1 = MontyObjectRecognitionExperiment()
-        with self.eval_exp_1:
-            self.eval_exp_1.setup_experiment(eval_cfg_1)
+        with MontyObjectRecognitionExperiment(eval_cfg_1) as self.eval_exp_1:
```
**question**: Is there a need for `eval_exp_1` to be attached to `self` here?

It looks like you didn't remove `self.` from anything with the `self.eval_exp_` pattern here and below.

tristanls commented on tests/unit/graph_learning_test.py:
```diff
@@ -1131,30 +1078,28 @@ def test_time_out(self):
+            exp.model.set_experiment_mode("train")
             pprint("...training...")
-            # self.exp.train()
+            # exp.train()
```
**issue**: Remove commented-out code.

tristanls commented on tests/unit/run_parallel_test.py:
```diff
@@ -241,10 +240,7 @@ def test_parallel_runs_n_epochs_lt(self):
-        with self.eval_exp:
-            self.eval_exp.setup_experiment(self.eval_config)
-
+        with MontyObjectRecognitionExperiment(self.eval_config) as self.eval_exp:
```
**question**: Is there a need to keep `self.eval_exp` here?

tristanls commented on tests/unit/run_parallel_test.py:
```diff
@@ -292,10 +288,7 @@ def test_parallel_runs_n_epochs_lt(self):
-        with self.eval_exp_lt:
-            self.eval_exp_lt.setup_experiment(self.eval_config_lt)
-
+        with MontyObjectRecognitionExperiment(self.eval_config_lt) as self.eval_exp_lt:
```
**question**: Is there a need to keep `self.eval_exp_lt` here?

tristanls reviewed (commented):
Thank you for the improvements. Readability is definitely improved as a result.

I think there are still a few omissions to address.

tristanls reviewed (approved):
Great, thank you for the remaining updates!

tristanls commented:
NOTE: I force-pushed as the result of updating the branch via a rebase.
